File: browser-runtime/src/quicker_browser_runtime/protocol.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class RefTarget:
    role: str
    name: str | None
    nth: int


def format_snapshot_yaml(
    url: str,
    title: str,
    ref_map: dict[str, RefTarget],
) -> str:
    lines = [
        f"url: {url}",
        f"title: {title}",
        "nodes:",
    ]
    for ref, target in ref_map.items():
        indent = "  "
        name_part = f' name="{target.name}"' if target.name else ""
        nth_part = f" nth={target.nth}" if target.nth > 0 else ""
        lines.append(
            f"{indent}- role={target.role}{name_part} ref={ref}{nth_part}",
        )
    return "\n".join(lines)


def parse_ref(ref: str) -> str:
    value = ref.strip()
    if not value:
        raise ValueError("ref is required")
    return value


def invoke_error(message: str, *, code: str = "error") -> dict[str, Any]:
    return {"ok": False, "error": code, "message": message}


def invoke_ok(data: dict[str, Any] | None = None) -> dict[str, Any]:
    payload: dict[str, Any] = {"ok": True}
    if data is not None:
        payload["data"] = data
    return payload

File: agent-gui/browser-runtime/protocol.mjs
/** @typedef {{ role: string; name: string | null; nth: number }} RefTarget */

/**
 * @param {string} url
 * @param {string} title
 * @param {Record<string, RefTarget>} refMap
 */
export function formatSnapshotYaml(url, title, refMap) {
  const lines = [`url: ${url}`, `title: ${title}`, "nodes:"];
  for (const [ref, target] of Object.entries(refMap)) {
    const namePart = target.name ? ` name="${target.name}"` : "";
    const nthPart = target.nth > 0 ? ` nth=${target.nth}` : "";
    const hrefPart =
      typeof target.href === "string" && target.href.trim()
        ? ` href=${target.href.trim()}`
        : "";
    lines.push(`  - role=${target.role}${namePart} ref=${ref}${nthPart}${hrefPart}`);
  }
  return lines.join("\n");
}

/** @param {string} message @param {{ code?: string }} [opts] */
export function invokeError(message, opts = {}) {
  return { ok: false, error: opts.code ?? "error", message };
}

/** @param {Record<string, unknown> | null | undefined} [data] */
export function invokeOk(data) {
  const payload = { ok: true };
  if (data != null) payload.data = data;
  return payload;
}
